File: web/src/lib/routing-sankey.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Dual-licensed: AGPL-3.0-or-later (open source) or commercial license (closed-source
// deployment exemption); see LICENSE and LICENSE.commercial. Copyright (c) 2026 is7Qin.

// 路由流 Sankey 纯数据构造器：服务端边 → (route → ordinal/lane →
// account → outcome/terminal) 四层节点 + 去重链路。铁律：不重算任何服务端数值
// （Wilson/frontier/plan/守恒判定都不在这里），只做拓扑映射与 chain_count 求和。
// 输出对同一输入字节级确定：节点按 (kind rank, 语义键) 排序，链路按 (source,target) 排序。
//
// 折叠语义（§4.2）：`folded` 是权威判别位——折叠边（后端 account_id=0 的「其他」
// 聚合）按 (ordinal, lane) 分层归入该层独立的「其他」节点（`other:<ordinal>:<lane>`），
// 不得按 account_id 跨层合并。previous_accounts 对折叠节点恒为空（后端即如此），
// 故折叠边的 tooltip 只挂 transition_reasons / previous_outcomes。

import type { components } from '@/lib/api/schema'

type GraphEdge = components['schemas']['RoutingFlowGraphEdge']

export type FlowSankeyNodeKind = 'route' | 'lane' | 'account' | 'outcome' | 'terminal'

export interface FlowSankeyNode {
  /** 稳定 ID：route / lane:<ordinal>:<lane> / acct:<id> / other:<ordinal>:<lane> / out:<outcome> / term:<outcome> */
  id: string
  /** 本地化显示名（Recharts nameKey 消费） */
  name: string
  kind: FlowSankeyNodeKind
}

export interface FlowSankeyLink {
  source: number
  target: number
  value: number
  sourceId: string
  targetId: string
  /** lane→account 到达链的 transition_reason 去重集（升序）；其余链为空 */
  reasons: string[]
  /** 到达链的 previous_outcome 去重集（升序） */
  prevOutcomes: string[]
  /** 到达链的 previous_account_id 去重集（升序）；折叠节点恒为空 */
  prevAccounts: number[]
  /** 到达链中是否含首发起边（previous_account_id=null） */
  hasFirstDispatch: boolean
}

export interface FlowSankeyData {
  nodes: FlowSankeyNode[]
  links: FlowSankeyLink[]
}

export interface FlowSankeyLabels {
  route: string
  lane: (ordinal: number, lane: string) => string
  account: (accountId: number) => string
  /** folded=true 的层独立「其他」聚合节点的聚合列标签 */
  foldedAccount: () => string
  outcome: (outcome: string) => string
  terminal: (outcome: string) => string
}

const KIND_RANK: Record<FlowSankeyNodeKind, number> = {
  route: 0,
  lane: 1,
  account: 2,
  outcome: 3,
  terminal: 4,
}

interface NodeDraft {
  id: string
  name: string
  kind: FlowSankeyNodeKind
  /** 列内排序键：lane=(ordinal,lane)；account=(id)；outcome/terminal=(outcome) */
  sortKey: readonly (string | number)[]
}

interface LinkDraft {
  sourceId: string
  targetId: string
  value: number
  reasons: Set<string>
  prevOutcomes: Set<string>
  prevAccounts: Set<number>
  hasFirstDispatch: boolean
}

/** 归一化后的边行：两种服务端边形（明细 / 桑基聚合）在此统一。 */
interface EdgeLike {
  ordinal: number
  lane: string
  /** 账号列节点稳定 ID：明细边为 acct:<id>；折叠边为层独立 other:<ordinal>:<lane> */
  accountNodeId: string
  accountName: string
  /** 层内排序键：明细边按 account_id；折叠边恒排该层末尾 */
  accountSortKey: readonly (string | number)[]
  outcome: string
  isTerminal: boolean
  chainCount: number
  reasons: readonly string[]
  prevOutcomes: readonly string[]
  prevAccounts: readonly number[]
  hasFirstDispatch: boolean
}

function buildSankey(likes: readonly EdgeLike[], labels: FlowSankeyLabels): FlowSankeyData {
  const nodes = new Map<string, NodeDraft>()
  const links = new Map<string, LinkDraft>()

  const ensureNode = (draft: NodeDraft) => {
    if (!nodes.has(draft.id)) nodes.set(draft.id, draft)
  }
  const addLink = (sourceId: string, targetId: string, like: EdgeLike, arrival: boolean) => {
    const key = `${sourceId}\u0000${targetId}`
    let draft = links.get(key)
    if (!draft) {
      draft = {
        sourceId,
        targetId,
        value: 0,
        reasons: new Set(),
        prevOutcomes: new Set(),
        prevAccounts: new Set(),
        hasFirstDispatch: false,
      }
      links.set(key, draft)
    }
    draft.value += like.chainCount
    if (!arrival) return
    // 迁移拓扑只挂在 lane→account 到达链上：reason/prev 描述「为什么走到这个账号」。
    for (const r of like.reasons) draft.reasons.add(r)
    for (const o of like.prevOutcomes) draft.prevOutcomes.add(o)
    for (const a of like.prevAccounts) draft.prevAccounts.add(a)
    if (like.hasFirstDispatch) draft.hasFirstDispatch = true
  }

  ensureNode({ id: 'route', name: labels.route, kind: 'route', sortKey: [] })
  for (const like of likes) {
    const laneId = `lane:${like.ordinal}:${like.lane}`
    const terminal = like.isTerminal
    const outcomeId = `${terminal ? 'term' : 'out'}:${like.outcome}`
    ensureNode({ id: laneId, name: labels.lane(like.ordinal, like.lane), kind: 'lane', sortKey: [like.ordinal, like.lane] })
    ensureNode({ id: like.accountNodeId, name: like.accountName, kind: 'account', sortKey: like.accountSortKey })
    ensureNode({
      id: outcomeId,
      name: terminal ? labels.terminal(like.outcome) : labels.outcome(like.outcome),
      kind: terminal ? 'terminal' : 'outcome',
      sortKey: [like.outcome],
    })
    addLink('route', laneId, like, false)
    addLink(laneId, like.accountNodeId, like, true)
    addLink(like.accountNodeId, outcomeId, like, false)
  }

  const sortedKeys = (a: readonly (string | number)[], b: readonly (string | number)[]) => {
    for (let i = 0; i < Math.min(a.length, b.length); i++) {
      const x = a[i]!
      const y = b[i]!
      const cmp = typeof x === 'number' && typeof y === 'number' ? x - y : String(x).localeCompare(String(y))
      if (cmp !== 0) return cmp
    }
    return a.length - b.length
  }
  const ordered = [...nodes.values()].sort(
    (a, b) => KIND_RANK[a.kind] - KIND_RANK[b.kind] || sortedKeys(a.sortKey, b.sortKey) || a.id.localeCompare(b.id)
  )
  const indexById = new Map(ordered.map((n, i) => [n.id, i]))

  const outLinks = [...links.values()]
    .map<FlowSankeyLink>(draft => ({
      source: indexById.get(draft.sourceId)!,
      target: indexById.get(draft.targetId)!,
      value: draft.value,
      sourceId: draft.sourceId,
      targetId: draft.targetId,
      reasons: [...draft.reasons].sort(),
      prevOutcomes: [...draft.prevOutcomes].sort(),
      prevAccounts: [...draft.prevAccounts].sort((x, y) => x - y),
      hasFirstDispatch: draft.hasFirstDispatch,
    }))
    .sort((a, b) => a.source - b.source || a.target - b.target)

  return {
    nodes: ordered.map(({ id, name, kind }) => ({ id, name, kind })),
    links: outLinks,
  }
}

/** 桑基聚合边（RoutingFlowGraphEdge）：folded 边按层归入独立「其他」节点。 */
export function buildFoldedFlowSankey(edges: readonly GraphEdge[], labels: FlowSankeyLabels): FlowSankeyData {
  return buildSankey(
    edges.map<EdgeLike>(edge => {
      if (!edge.folded) {
        return {
          ordinal: edge.ordinal,
          lane: edge.lane,
          accountNodeId: `acct:${edge.account_id}`,
          accountName: labels.account(edge.account_id),
          accountSortKey: [0, edge.account_id],
          outcome: edge.outcome,
          isTerminal: edge.is_terminal,
          chainCount: edge.chain_count,
          reasons: edge.transition_reasons,
          prevOutcomes: edge.previous_outcomes,
          prevAccounts: edge.previous_accounts,
          hasFirstDispatch: false,
        }
      }
      // 折叠边：以前缀 other 按层独立成节点（不跨层合并）；previous_accounts
      // 后端恒为空，此处直接沿用（不推导）；transition_reasons/previous_outcomes 照常挂载。
      return {
        ordinal: edge.ordinal,
        lane: edge.lane,
        accountNodeId: `other:${edge.ordinal}:${edge.lane}`,
        accountName: labels.foldedAccount(),
        accountSortKey: [1, ''],
        outcome: edge.outcome,
        isTerminal: edge.is_terminal,
        chainCount: edge.chain_count,
        reasons: edge.transition_reasons,
        prevOutcomes: edge.previous_outcomes,
        prevAccounts: [],
        hasFirstDispatch: false,
      }
    }),
    labels,
  )
}

File: internal/service/routing_sankey.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Dual-licensed: AGPL-3.0-or-later (open source) or commercial license (closed-source
// deployment exemption); see LICENSE and LICENSE.commercial. Copyright (c) 2026 is7Qin.

package service

// 路由 flow 的桑基图边集构造：每 (ordinal, lane) 层按 chain_count 保留 top-N
// 账号，其余折叠为 account_id=0 的「其他」节点。
//
// 两条不变量：
//  1. 折叠只合并账号，层内 chain_count 求和不变——守恒不破，图形不因折叠失真。
//  2. sankey 恒在**完整边集**上折叠，与边表分页（offset/limit）严格解耦——
//     翻页不改变图形，图形也不随翻页变化。
//
// 折叠是纯拓扑操作，不重算任何服务端统计量（Wilson/frontier/守恒判定都不在此）。

import "sort"

// RoutingFlowGraphEdge 一条桑基边（已按 (ordinal, lane, account, outcome,
// is_terminal) 聚合）。Folded=true 表示「其他」聚合节点（AccountID 恒 0）。
type RoutingFlowGraphEdge struct {
	Ordinal    int16
	Lane       string
	AccountID  int64
	Folded     bool
	Outcome    string
	IsTerminal bool
	ChainCount int64
	// 到达链去重集（升序）。折叠节点不收集 PreviousAccounts——折叠合并了多个
	// 账号，前驱集合会膨胀且无展示价值；reasons/outcomes 是小枚举，照常收集。
	TransitionReasons []string
	PreviousOutcomes  []string
	PreviousAccounts  []int64
}

// RoutingFlowGraph 桑基图边集 + 折叠元数据。
type RoutingFlowGraph struct {
	Edges          []RoutingFlowGraphEdge
	AccountLimit   int
	FoldedAccounts int64
	Folded         bool
}

// sankeyLayerKey 折叠分层的身份：(ordinal, lane)。
type sankeyLayerKey struct {
	ordinal int16
	lane    string
}

// sankeyLayerAccount 层内账号身份。
type sankeyLayerAccount struct {
	layer   sankeyLayerKey
	account int64
}

// sankeyEdgeKey 聚合身份（不含 generation/fingerprint——桑基不消费这两个维度，
// 它们正是边数膨胀的来源，聚合掉才能让图形有界）。
type sankeyEdgeKey struct {
	ordinal    int16
	lane       string
	accountID  int64
	outcome    string
	isTerminal bool
}

// buildFlowSankey 把完整边集折叠成有界桑基边集。
//
// accountLimit ≤0 视为不折叠（保留全部账号）；由调用方按契约归一后再传入。
// 输出对同一输入字节级确定：层内账号按 (chain_count 降序, account_id 升序)
// 排名，边按 (ordinal, lane, account_id, outcome, is_terminal) 升序。
func buildFlowSankey(edges []RoutingFlowEdge, accountLimit int) RoutingFlowGraph {
	// 1. 层内账号的 chain_count 合计（排名依据）。
	layerTotals := make(map[sankeyLayerAccount]int64)
	layerAccounts := make(map[sankeyLayerKey]map[int64]struct{})
	for _, e := range edges {
		lk := sankeyLayerKey{e.Ordinal, e.Lane}
		la := sankeyLayerAccount{layer: lk, account: e.AccountID}
		layerTotals[la] += e.ChainCount
		if layerAccounts[lk] == nil {
			layerAccounts[lk] = make(map[int64]struct{})
		}
		layerAccounts[lk][e.AccountID] = struct{}{}
	}

	// 2. 每层独立排名，取 top-N 保留。
	kept := make(map[sankeyLayerAccount]bool)
	var foldedAccounts int64
	layerKeys := make([]sankeyLayerKey, 0, len(layerAccounts))
	for lk := range layerAccounts {
		layerKeys = append(layerKeys, lk)
	}
	sort.Slice(layerKeys, func(i, j int) bool {
		if layerKeys[i].ordinal != layerKeys[j].ordinal {
			return layerKeys[i].ordinal < layerKeys[j].ordinal
		}
		return layerKeys[i].lane < layerKeys[j].lane
	})
	for _, lk := range layerKeys {
		ids := make([]int64, 0, len(layerAccounts[lk]))
		for id := range layerAccounts[lk] {
			ids = append(ids, id)
		}
		sort.Slice(ids, func(i, j int) bool {
			ai := layerTotals[sankeyLayerAccount{layer: lk, account: ids[i]}]
			aj := layerTotals[sankeyLayerAccount{layer: lk, account: ids[j]}]
			if ai != aj {
				return ai > aj // 链数多的优先
			}
			return ids[i] < ids[j] // 同链数按账号升序（确定性）
		})
		for i, id := range ids {
			if accountLimit > 0 && i >= accountLimit {
				foldedAccounts++
				continue
			}
			kept[sankeyLayerAccount{layer: lk, account: id}] = true
		}
	}

	// 3. 按 (ordinal, lane, account, outcome, is_terminal) 聚合；未保留账号归入占位
	// id 并置 folded=true。folded 是权威判别位（I6）：不得以 account_id == 0
	// 推断——占位 id 恰为 0，消费方必须以 folded 为准。
	type agg struct {
		chainCount int64
		folded     bool
		reasons    map[string]struct{}
		prevOut    map[string]struct{}
		prevAccts  map[int64]struct{}
	}
	aggs := make(map[sankeyEdgeKey]*agg)
	for _, e := range edges {
		lk := sankeyLayerKey{e.Ordinal, e.Lane}
		isKept := kept[sankeyLayerAccount{layer: lk, account: e.AccountID}]
		acct := e.AccountID
		if !isKept {
			acct = 0
		}
		k := sankeyEdgeKey{e.Ordinal, e.Lane, acct, e.Outcome, e.IsTerminal}
		a := aggs[k]
		if a == nil {
			a = &agg{
				folded:    !isKept,
				reasons:   make(map[string]struct{}),
				prevOut:   make(map[string]struct{}),
				prevAccts: make(map[int64]struct{}),
			}
			aggs[k] = a
		}
		a.chainCount += e.ChainCount
		if e.TransitionReason != "" {
			a.reasons[e.TransitionReason] = struct{}{}
		}
		if e.PreviousOutcome != "" {
			a.prevOut[e.PreviousOutcome] = struct{}{}
		}
		if isKept && e.PreviousAccountID != nil {
			a.prevAccts[*e.PreviousAccountID] = struct{}{}
		}
	}

	// 4. 确定性输出序。
	keys := make([]sankeyEdgeKey, 0, len(aggs))
	for k := range aggs {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool {
		a, b := keys[i], keys[j]
		if a.ordinal != b.ordinal {
			return a.ordinal < b.ordinal
		}
		if a.lane != b.lane {
			return a.lane < b.lane
		}
		if a.accountID != b.accountID {
			return a.accountID < b.accountID
		}
		if a.outcome != b.outcome {
			return a.outcome < b.outcome
		}
		return !a.isTerminal && b.isTerminal
	})

	out := make([]RoutingFlowGraphEdge, 0, len(keys))
	for _, k := range keys {
		a := aggs[k]
		out = append(out, RoutingFlowGraphEdge{
			Ordinal:           k.ordinal,
			Lane:              k.lane,
			AccountID:         k.accountID,
			Folded:            a.folded,
			Outcome:           k.outcome,
			IsTerminal:        k.isTerminal,
			ChainCount:        a.chainCount,
			TransitionReasons: sortedStringSet(a.reasons),
			PreviousOutcomes:  sortedStringSet(a.prevOut),
			PreviousAccounts:  sortedInt64Set(a.prevAccts),
		})
	}
	return RoutingFlowGraph{
		Edges:          out,
		AccountLimit:   accountLimit,
		FoldedAccounts: foldedAccounts,
		Folded:         foldedAccounts > 0,
	}
}

func sortedStringSet(m map[string]struct{}) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

func sortedInt64Set(m map[int64]struct{}) []int64 {
	out := make([]int64, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Slice(out, func(i, j int) bool { return out[i] < out[j] })
	return out
}
